examples/sparse/graphsage.py

```diff
@@ -0,0 +1,230 @@
+        # Divided by degree.
+        D_hat = dglsp.diag(A.sum(0)) ** -1
+        A_div = A @ D_hat
+        # Conv neighbors.
+        dstdata = A_div.T @ srcdata
+
+        rst = self.fc_self(feat_dst) + dstdata
+        return rst
+
+
+class SAGE(nn.Module):
+    def __init__(self, in_size, hid_size, out_size):
+        super().__init__()
+        self.layers = nn.ModuleList()
+        # Three-layer GraphSAGE-gcn.
+        self.layers.append(SAGEConv(in_size, hid_size))
+        self.layers.append(SAGEConv(hid_size, hid_size))
+        self.layers.append(SAGEConv(hid_size, out_size))
+        self.dropout = nn.Dropout(0.5)
+        self.hid_size = hid_size
+        self.out_size = out_size
+
+    def forward(self, sampled_mats, x):
+        hidden_x = x
+        for layer_idx, (layer, sampled_mat) in enumerate(
+            zip(self.layers, sampled_mats)
+        ):
+            hidden_x = layer(sampled_mat, hidden_x)
+            if layer_idx != len(self.layers) - 1:
+                hidden_x = F.relu(hidden_x)
+                hidden_x = self.dropout(hidden_x)
+        return hidden_x
+
+
+def evaluate(model, A, dataloader, ndata, num_classes):
+    model.eval()
+    ys = []
+    y_hats = []
+    fanouts = [10, 10, 10]
+    for it, (seeds) in enumerate(dataloader):
+        with torch.no_grad():
+            src = seeds
+            sampled_mats = []
+            for fanout in fanouts:
+                # Sampling neighbor
+                sampled_mat = A.sample(1, fanout, ids=src, replace=True)
+                # Compact the matrix
+                compacted_mat, row_ids = sampled_mat.compact(0)
+                sampled_mats.insert(0, compacted_mat)
+                src = row_ids
+
+            x = ndata["feat"][src]
+            y = ndata["label"][seeds]
+            ys.append(y)
+            y_hats.append(model(sampled_mats, x))
+
+    return MF.accuracy(
+        torch.cat(y_hats),
+        torch.cat(ys),
+        task="multiclass",
+        num_classes=num_classes,
+    )
+
+
+def validate(device, A, ndata, dataset, model, batch_size):
+    model.eval()
+    inf_id = dataset.test_idx.to(device)
+    inf_dataloader = torch.utils.data.DataLoader(inf_id, batch_size=batch_size)
+    acc = evaluate(model, A, inf_dataloader, ndata, dataset.num_classes)
+    return acc
+
+
+def train(device, A, ndata, dataset, model):
+    # Create sampler & dataloader.
+    train_idx = dataset.train_idx.to(device)
+    val_idx = dataset.val_idx.to(device)
+
+    train_dataloader = torch.utils.data.DataLoader(
+        train_idx, batch_size=1024, shuffle=True
+    )
+    val_dataloader = torch.utils.data.DataLoader(val_idx, batch_size=1024)
+
+    opt = torch.optim.Adam(model.parameters(), lr=1e-3, weight_decay=5e-4)
+
+    fanouts = [10, 10, 10]
+    for epoch in range(10):
+        model.train()
+        total_loss = 0
+        for it, (seeds) in enumerate(train_dataloader):
+            src = seeds
+            sampled_mats = []
```

Comment:
The the train / validation and test code are quite similar from iterate the dataloader to generate sampled_mats and x, y. consider refactor to simplify the code.